source/client_channel_handler.c

```diff
@@ -531,103 +595,15 @@ static int s_process_read_message(
-        bool packet_complete = true;
-        if (to_read > message_cursor.len) {
-            to_read = message_cursor.len;
-            packet_complete = false;
-        }
-
-        /* Write the chunk to the buffer.
-         * This will either complete the packet, or be the entirety of message if more data is required. */
-        struct aws_byte_cursor chunk = aws_byte_cursor_advance(&message_cursor, to_read);
-        AWS_ASSERT(chunk.ptr); /* Guaranteed to be in bounds */
-        result = (int)aws_byte_buf_write_from_whole_cursor(&connection->thread_data.pending_packet, chunk) - 1;
-        if (result) {
-            goto handle_error;
-        }
-
-        /* If the packet is still incomplete, don't do anything with the data. */
-        if (!packet_complete) {
-            AWS_LOGF_TRACE(
-                AWS_LS_MQTT_CLIENT,
-                "id=%p: partial message is still incomplete, waiting on another read.",
-                (void *)connection);
-
-            goto cleanup;
-        }
-
-        /* Handle the completed pending packet */
-        struct aws_byte_cursor packet_data = aws_byte_cursor_from_buf(&connection->thread_data.pending_packet);
-        AWS_LOGF_TRACE(AWS_LS_MQTT_CLIENT, "id=%p: full mqtt packet re-assembled, dispatching.", (void *)connection);
-        result = s_process_mqtt_packet(connection, aws_mqtt_get_packet_type(packet_data.ptr), packet_data);
-
-    handle_error:
-        /* Clean up the pending packet */
-        aws_byte_buf_clean_up(&connection->thread_data.pending_packet);
-        AWS_ZERO_STRUCT(connection->thread_data.pending_packet);
-
-        if (result) {
-            return AWS_OP_ERR;
-        }
-    }
-
-    while (message_cursor.len) {
-
-        /* Temp byte cursor so we can decode the header without advancing message_cursor. */
-        struct aws_byte_cursor header_decode = message_cursor;
-
-        struct aws_mqtt_fixed_header packet_header;
-        AWS_ZERO_STRUCT(packet_header);
-        int result = aws_mqtt_fixed_header_decode(&header_decode, &packet_header);
-
-        /* Calculate how much data was read. */
-        const size_t fixed_header_size = message_cursor.len - header_decode.len;
-
-        if (result) {
-            if (aws_last_error() == AWS_ERROR_SHORT_BUFFER) {
-                /* Message data too short, store data and come back later. */
-                AWS_LOGF_TRACE(
-                    AWS_LS_MQTT_CLIENT, "id=%p: message is incomplete, waiting on another read.", (void *)connection);
-                if (aws_byte_buf_init(
-                        &connection->thread_data.pending_packet,
-                        connection->allocator,
-                        fixed_header_size + packet_header.remaining_length)) {
-
-                    return AWS_OP_ERR;
-                }
+    int result = aws_mqtt311_decoder_on_bytes_received(&connection->thread_data.decoder, message_cursor);
```

Comment:
might want to take more drastic action after a decoder failure, like putting the connection into a state where it refuses to process any more data. or have aws_mqtt311_decoder_on_bytes_received() first check that it's not already in an error state

aws_channel_shutdown() isn't immediate. Other channel handlers might have data queued up that they'll keep sending the the MQTT-handler. It would be pretty bad to report a bad packet, but then continue processing the next

Reply:
Well once the decoder goes into a terminal state, it just always returns errors on incoming data

Reply:
currently, if you call aws_mqtt311_decoder_on_bytes_received() again after an error, it still processes the new data

Reply:
Once it's in the protocol error state it just returns AWS_OP_ERR to subsequent calls.